feat(linter): eslint/sort-keys

Goldziher opened the pull request:
This PR adds the `eslint/sort-keys` implementation as part of #479 

Note: two tests are commented out. I was not able to model this behavior properly. I would be happy to have some assistance there.

Goldziher committed:
chore: added scaffolding for eslint/sort-keys

Goldziher committed:
feat: sorts-keys initial implementation

Goldziher committed:
feat: add sorting logic

Goldziher committed:
feat: almost passing tests

Goldziher committed:
Merge branch 'oxc-project:main' into main

DonIsaac commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1077 @@
+            let is_sorted = all(property_groups.iter().zip(&sorted_property_groups), |(a, b)| a == b);
+
+            if !is_sorted {
+                ctx.diagnostic(
```
Use a diagnostic factory function. See other rules for examples.

Goldziher replied:
👍🏼

DonIsaac commented:
Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

Goldziher commented on .gitignore:
```diff
@@ -27,3 +27,5 @@ tasks/coverage/babel/
 tasks/coverage/typescript/
 tasks/prettier_conformance/prettier/
+
+.idea/
```
Ok, why?

Goldziher replied:
done. 

I simply don't commit it - but I do commit it to other projects, which are not open source, so I am not globally ignoring it. I would recommend its ignored in OXC instead.

rzvxa replied:
Feel free to mark this conversation as resolved, We can add it in a follow-up if Boshen gives a 👍

@DonIsaac

DonIsaac replied:
Boshen has expressed that editor-specific files should not be included in oxc's .gitignore. I don't recall his reasoning, but I know he doesn't want it there.

Goldziher commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
````diff
@@ -0,0 +1,1077 @@
+    /// FIXME: Tests will fail if examples are missing or syntactically incorrect.
+    /// ```
+    SortKeys,
+    nursery, // TODO: change category to `correctness`, `suspicious`, `pedantic`, `perf`, `restriction`, or `style`
````
Ok. Which one would this rule be? Style?

Goldziher replied:
im proceeding for now with sytle. Lemme know if something else is better.

Goldziher commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1077 @@
+
+impl Rule for SortKeys {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config_array = match value.as_array() {
```
👍🏼

Goldziher commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1077 @@
+            Some(v) => v,
+            None => {
+                // we follow the eslint defaults
+                return Self(Box::new(SortKeysOptions {
```
👍🏼

Goldziher commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
````diff
@@ -0,0 +1,1077 @@
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// FIXME: Tests will fail if examples are missing or syntactically incorrect.
````
👍🏼

Goldziher committed:
Merge branch 'main' into main

Goldziher committed:
chore: addressed review comments

Goldziher committed:
Merge branch 'main' of github.com:Goldziher/oxc

Goldziher committed:
chore: update doc string

Goldziher commented:
> Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

Great. I addressed all comments.

DonIsaac commented on Cargo.toml:
```diff
@@ -192,7 +192,7 @@ ignored = ["napi", "oxc_transform_napi", "prettyplease"]
 # Disabling debug info speeds up local and CI builds,
 # and we don't rely on it for debugging that much.
-debug = false
+# debug = false
```
Please undo this

Goldziher replied:
```suggestion
```

Goldziher replied:
unintentional. Reverted. Note- this is an issue for running a debugger in intellij.

DonIsaac replied:
Yup, same here for me in VSCode. I simply don't commit it; it helps keep CI nice and speedy 🏃

Goldziher committed:
Update Cargo.toml

Goldziher committed:
chore: revert unintentional change to cargo.toml

Goldziher committed:
Merge branch 'main' into main

DonIsaac committed:
Merge branch 'main' into main

DonIsaac commented:
@rzvxa feel free to merge this when CI is passing and you are ready.

Goldziher committed:
Merge branch 'oxc-project:main' into main

Goldziher committed:
chore: fix failing snapshot

DonIsaac committed:
Merge branch 'main' into main

rzvxa commented:
@Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?

To fix the CI issues you can run `just r` on your local machine.
For updating snapshots you should do the following:

1. `cargo insta test`
2. `cargo insta review`
3. accept or reject snapshot updates
3. commit the updated snapshots.

The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.

The one with `new` in the name is for local use in the review process and shouldn't be committed.

Goldziher committed:
chore: run just r

Goldziher committed:
Merge branch 'main' into main

Goldziher committed:
Merge branch 'main' of github.com:Goldziher/oxc

Goldziher commented:
> @Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?
> 
> To fix the CI issues you can run `just r` on your local machine. For updating snapshots you should do the following:
> 
> 1. `cargo insta test`
> 2. `cargo insta review`
> 3. accept or reject snapshot updates
> 4. commit the updated snapshots.
> 
> The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.
> 
> The one with `new` in the name is for local use in the review process and shouldn't be committed.

thanks, its done.

rzvxa commented:
There are still failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

Goldziher commented:
> There are still some failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

I'll try running it tomorrow, cheers

Goldziher committed:
Merge branch 'main' into main

Goldziher committed:
chore: fixed clippy issues

Goldziher committed:
chore: added semi-columns to examples

Goldziher commented:
Hi, @rzvxa - i fixed all the clippy issues and ensured the tests are all passing. I am getting one error running `just r` that I don't quite understand:

```text
est linter::rules::test::test_doc_pages ... FAILED
test linter::rules::test::test_rules_table ... ok

failures:

---- linter::rules::test::test_doc_pages stdout ----
  x Expected a semicolon or an implicit semicolon after a statement, but found
  | none
   ,-[Pedantic/eslint/sort-keys/code-snippet:1:10]
 1 | let myObj {
   :          ^
 2 |   c: 1,
   `----
  help: Try insert a semicolon here

thread 'linter::rules::test::test_doc_pages' panicked at tasks/website/src/linter/rules/test.rs:94:64:
called `Result::unwrap()` on an `Err` value: "1 errors occurred while parsing Pedantic/eslint/sort-keys/code-snippet.jsx"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    linter::rules::test::test_doc_pages

test result: FAILED. 5 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s

error: test failed, to rerun pass `-p website --lib`
error: Recipe `test` failed on line 67 with exit code 101
error: Recipe `ready` failed on line 27 with exit code 101
```

I don't understand what semi-colon it wants from me to add - the JS is find and there is a semi-colon at the end of the statement.

DonIsaac committed:
Merge branch 'main' into main

rzvxa commented:
You are missing an equal sign there:

```js
let obj = { /* ... */ }
```
Since there is no equal sign it interprets it as two expressions e.g.
```js
let obj; { /* ... */ }
```
And says to add a semicolon to be more explicit.

Goldziher committed:
chore: fix missing equal signs

Goldziher committed:
Merge branch 'main' of github.com:Goldziher/oxc

Goldziher committed:
Merge branch 'main' into main

Goldziher commented:
> You are missing an equal sign there:
> 
> ```js
> let obj = { /* ... */ }
> ```
> 
> Since there is no equal sign it interprets it as two expressions e.g.
> 
> ```js
> let obj; { /* ... */ }
> ```
> 
> And says to add a semicolon to be more explicit.

🤦🏼 

All fixed and everything is green locally. Thanks for your help!

jelly commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1093 @@
+
+    pending  // TODO: describe fix capabilities. Remove if no fix can be done,
+             // keep at 'pending' if you think one could be added but don't know how.
+             // Options are 'fix', 'fix_dangerous', 'suggestion', and 'conditional_fix_suggestion'
```
The TODO comments can be removed as they placeholder has already been replaced.

Goldziher replied:
fixed

Goldziher committed:
Merge branch 'main' into main

Goldziher committed:
chore: removed todo comments

Goldziher committed:
Merge branch 'main' of github.com:Goldziher/oxc

Goldziher commented:
@rzvxa is this PR done?

DonIsaac commented:
@rzvxa appears busy, I'll review this in the morning.

DonIsaac committed:
Merge branch 'main' into main

Goldziher commented:
> @rzvxa appears busy, I'll review this in the morning. 

Anything?

DonIsaac committed:
Merge branch 'main' into main

DonIsaac commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1089 @@
+impl Rule for SortKeys {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let Some(config_array) = value.as_array() else {
+            return Self(Box::default());
```
Consider `Self::default()` for clarity

Goldziher replied:
done

DonIsaac commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1089 @@
+    }
+
+    let len = a.len().min(b.len());
+    let a_chars: Vec<char> = a.chars().take(len).collect();
```
Can you zip these two iterations together without collecting?

Goldziher replied:
done

DonIsaac commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1089 @@
+        return a_char.cmp(b_char);
+    }
+
+    a.cmp(b)
```
Can we not return `Ordering::Equal` here?

Goldziher replied:
I am not 100% sure, thats why I kept it like this.

DonIsaac commented on crates/oxc_linter/src/rules/eslint/sort_keys.rs:
```diff
@@ -0,0 +1,1089 @@
+        ("var obj = {1:1, '11':2, 2:4, A:3}", Some(serde_json::json!([]))),
+        ("var obj = {'#':1, 'Z':2, À:3, è:4}", Some(serde_json::json!([]))),
+        ("var obj = { [/(?<zero>0)/]: 1, '/(?<zero>0)/': 2 }", Some(serde_json::json!([]))), // { "ecmaVersion": 2018 },
+        // ("var obj = {a:1, b:3, [a + b]: -1, c:2}", Some(serde_json::json!([]))), // { "ecmaVersion": 6 },
```
What happened here?

Goldziher replied:
see the PR description - there are two cases where I wasn't able to match the test cases.

DonIsaac commented:
Apologies for the delay, I got sidetracked 😅

Goldziher committed:
chore: address review comments

Goldziher commented:
@DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

DonIsaac commented:
> @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

The conflicts aren't too bad, I believe in you 😄

Goldziher commented:
> > @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.
> 
> The conflicts aren't too bad, I believe in you 😄 

That's not the issue. Ill need to add a second remote now, but I'll do it.

Goldziher committed:
Merge branch 'main' of github.com:oxc-project/oxc

Goldziher commented:
Should be good to go @DonIsaac

DonIsaac commented:
Once CI is passing we can merge this

Goldziher committed:
chore: synced with upstream

Goldziher committed:
chore: ignore clippy issue

Goldziher committed:
chore: applied just r

DonIsaac committed:
Merge branch 'main' into main
